packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs

```diff
@@ -294,6 +999,521 @@ private MethodBodyStatement GetPropertyInitializers(IReadOnlyList<ParameterProvi
+                {
+                    ThrowValidationFailException(format, iModelTInterface.Arguments[0], validationType)
+                },
+                EmptyStatement
+            ];
+
+            return statements;
+        }
+
+        private MethodBodyStatement GetConcreteFormat(ModelReaderWriterOptionsSnippet options, CSharpType iModelTInterface, out TypedSnippet format)
+        {
+            var castSnippet = new StringSnippet(This.CastTo(iModelTInterface).Invoke(nameof(IPersistableModel<object>.GetFormatFromOptions), options));
+            var condition = new TernaryConditionalExpression(
+                Equal(options.Format, ModelReaderWriterOptionsSnippet.WireFormat),
+                castSnippet,
+                options.Format);
+            var reference = new VariableReferenceSnippet(castSnippet.Type, "format");
+            format = reference;
+            return Var(reference, condition);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="KeywordStatement"/> of type <see cref="FormatException"/> with a specific message indicating
+        /// that the model does not support the specified serialization format.
+        /// </summary>
+        /// <param name="format">The serialization format.</param>
+        /// <param name="typeOfT">The type of the model.</param>
+        /// <param name="validationType">The type of validation (write or read).</param>
+        /// <returns>The <see cref="MethodBodyStatement"/> representing the throw statement.</returns>
+        private KeywordStatement ThrowValidationFailException(ValueExpression format, CSharpType typeOfT, SerializationFormatValidationType validationType)
+            => Throw(New.Instance(
+                typeof(FormatException),
+                new FormattableStringExpression($"The model {{{0}}} does not support {(validationType == SerializationFormatValidationType.Write ? "writing" : "reading")} '{{{1}}}' format.",
+                [
+                    Nameof(typeOfT),
+                    format
+                ])));
+
+        /// <summary>
+        /// Constructs the body statements containing the serialization statements for the model properties.
+        /// </summary>
+        /// <param name="utf8JsonWriter">The Utf8JsonWriter to write the properties to.</param>
+        /// <param name="options">The model reader/writer options.</param>
+        private MethodBodyStatement[] CreateWritePropertiesStatements(Utf8JsonWriterSnippet utf8JsonWriter, ModelReaderWriterOptionsSnippet options)
+        {
+            var propertyCount = PropertySerializations.Length;
+            var propertyStatements = new MethodBodyStatement[propertyCount];
+            for (var i = 0; i < propertyCount; i++)
+            {
+                var inputProperty = _inputModel.Properties[i];
+                var propertyMember = new MemberExpression(null, PropertySerializations[i].Name);
```

Comment:
How is it guaranteed that the inputModel.Properties list and the PropertySerializations list will have the exact same length and order?